soc/xtensa/intel_adsp/ace_v1x/power.c

```diff
@@ -26,29 +26,194 @@
+{
+	uint32_t intenable;
+
+	__asm__ volatile("rsr.intenable %0" : "=r"(intenable));
+	return intenable;
+}
+
+static ALWAYS_INLINE uint32_t read_threadptr(void)
+{
+	uint32_t threadptr;
+
+	__asm__ volatile(
+		"rur.threadptr %0" : "=a"(threadptr));
+	return threadptr;
+}
+
+static ALWAYS_INLINE void write_threadptr(uint32_t threadptr)
+{
+	__asm__ volatile(
+		"wur.threadptr %0" : : "a" (threadptr) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_vecbase(void)
+{
+	uint32_t vecbase;
+
+	__asm__ volatile("rsr.vecbase %0"
+		: "=a"(vecbase));
+	return vecbase;
+}
+
+static ALWAYS_INLINE void write_vecbase(uint32_t vecbase)
+{
+	__asm__ volatile("wsr.vecbase %0"
+		: : "a"(vecbase) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave2(void)
+{
+	uint32_t excsave2;
+
+	__asm__ __volatile__("rsr.excsave2 %0"
+		: "=a"(excsave2));
+	return excsave2;
+}
+
+static ALWAYS_INLINE void write_excsave2(uint32_t excsave2)
+{
+	__asm__ __volatile__("wsr.excsave2 %0"
+		: : "a"(excsave2) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave3(void)
+{
+	uint32_t excsave3;
+
+	__asm__ __volatile__("rsr.excsave3 %0"
+		: "=a"(excsave3));
+	return excsave3;
+}
+
+static ALWAYS_INLINE void write_excsave3(uint32_t excsave3)
+{
+	__asm__ __volatile__("wsr.excsave3 %0"
+		: : "a"(excsave3) : "memory");
+}
+
+#define BITMASK_FROM_RANGE_UINT32(from, to) ((0xFFFFFFFF << (from)) & \
+					    (0xFFFFFFFF >> (sizeof(uint32_t) * 8 - 1 - (to))))
+#define LPSCTL_BATTR_MASK                   BITMASK_FROM_RANGE_UINT32(12, 16)
+#define LPSRAM_MAGIC_VALUE 0x13579BDF
+
+struct lpsram_header {
+	uint32_t alt_reset_vector;
+	uint32_t adsp_lpsram_magic;
+	void *lp_restore_vector;
+	uint32_t reserved;
+	uint32_t slave_core_vector;
+	uint8_t rom_bypass_vectors_reserved[0xC00 - 0x14];
+};
+
+enum PowerFlow {
+	D0_TO_D0i3,
+	D0i3_TO_D0
+};
+
+struct core_state {
+	enum PowerFlow flow;
+	uint32_t intenable;
+	uint32_t thread_ptr;
+	uint32_t vecbase;
+	uint32_t excsave2;
+	uint32_t excsave3;
+};
+
+static struct core_state core_desc[CONFIG_MP_NUM_CPUS] = { 0 };
+
+void dsp_pg_restore(void);
+
+__asm__("dsp_pg_restore:\n\t"
+	"  movi  a0, 0\n\t"
+	"  movi  a1, 1\n\t"
+	"  wsr   a1, WINDOWSTART\n\t"
+	"  wsr   a0, WINDOWBASE\n\t"
+	"  rsync\n\t"
+	"  movi a3, 0x20\n\t"
+	"  rsr a4, intenable\n\t"
+	"  or a4, a4, a3\n\t"
+	"  wsr a4, intenable\n\t"
+	"  rsync\n\t"
+	"  wsr a3, interrupt\n\t"
+	"  rsync\n\t"
+	"  loop1:\n\t"
+	"    nop\n\t"
+	"    j loop1\n\t");
+
+#define SOFTWARE_INTERRUPT_NUMBER 5
+
+void software_irq_handler(void *arg)
+{
+	uint32_t core_id = processor_id_reg();
+
+	if (core_desc[core_id].flow == D0_TO_D0i3) {
+		struct lpsram_header *lpsheader =
+			(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
+		lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
+		lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
+		core_desc[core_id].flow = D0i3_TO_D0;
+		z_xtensa_cache_flush_inv_all();
+		k_cpu_idle();
```

Comment:
I am really confuse with this code, isn't this isr used to wake up a core ? Why this is being used when the core is going to be idle ? Shouldn't this be handled in the `power_gate()` function ?

How this is supposed to be used ? When this interruption is triggered ?

Reply:
The idea behind isr was that we trigger software interrupt to force context switch and that way save core state into the memory. But for this moment I have gave it up. Maybe I'll come back to it later.